### ISSUE TITLE
Consider a different browser embed method

### ISSUE THREAD

**Issue by matthew-dean** [less/less.js#2521]
One of the things that's been around for a long time has been this way to attach Less stylesheets in-browser.

```
<link rel="stylesheet/less" type="text/css" href="styles.less" />
```

When I was doing some performance testing recently, I noticed two events which are created unnecessarily by this method. The first is obvious. It downloads styles.less but doesn't do anything with it, so there are two initial "requests" for styles.less, but theoretically, the second should be loaded from cache. (Hopefully?)

The second made me say, "Hmm." By telling the browser that this is a stylesheet, we're telling the browser to go ahead and try to parse it and pull out any relevant styles. Sure enough, Chrome showed an event of parsing the stylesheet, which is not really what we want. I assume we never see those styles because it never actually applies them based on the `rel` attribute, but it does parse them.

Now, I don't know how much of a performance hit these two things might be. BUT, considering that there are lots of other ways to indicate to less.js what the stylesheet is, then it seems like an unnecessary burden.

It's possible that even eliminating `type="text/css"` may do it, or there may be other link types that don't have the browser request (or parse) the resource.

**Comment by calvinjuarez** [comment]
`type="text/plain"` would be valid, though not very semantic. `type="text/less"` is not valid, and maybe a little presumptuous, but it's semantic.

As for the `rel` attribute, MDN has a nice [list of kosher rel values](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types "Link types - HTML (HyperText Markup Language) | MDN"), though none of those seem super helpful.

And here's a link to [the `<link>` attribute on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link "<link> - HTML (HyperText Markup Language) | MDN"), for possible solutions using other attributes.

**Comment by matthew-dean** [comment]
Just a note about this: reading some stuff about preloading prioritization, and adding ```type="text/css"``` definitely places the .less download at the highest priority for some browsers.

**Comment by rjgotten** [comment]
> Hmm. I wonder if we could get the browser to download but then ignore / prevent parsing. 

`<a link="stylesheet alternate">` perhaps?

**Comment by seven-phases-max** [comment]
><link rel="stylesheet alternate prefetch">  perhaps?

Is `prefetch` supported by any browser?
And `alternate` seems to mean quite different thing (it's more like a hack for switching themes stuff and does not really related to the link *type*).

**Comment by rjgotten** [comment]
> Is prefetch supported by any browser?

Firefox Chrome and recent IE browsers implement it.

> (it's more like a hack for switching themes stuff and it's not really related to the link type)

I'm not 100% sure, but iirc the "alternate" type doesn't parse the downloaded CSS until it actually comes into use. So in this case it would serve to prevent the CSS parser from doing needless work on Less files.


If you want to have something more robust, but with more limited browser support:
`<link rel="subresource" href="styles.less" />`

This will flag a resource as critical for the current page load, will give them the absolute highest download priority and fetch them immediately. In this particular case, the CSS parser should not be invoked on the Less resource as the link's relation does not specify "stylesheet".

**Comment by matthew-dean** [comment]
> Firefox Chrome and recent IE browsers implement it, but browsers schedule its priority differently. Iirc it's scheduled quite low on Firefox even. 

That's right. Prefetch is basically "grab this resource when it's convenient". Usually that happens right after or around the onload event. So that's not what we want. "subresource" WOULD probably give us what we want (although still a lower priority than text/css), but I think is only supported by Chrome.

> Honestly, I feel this entire exercise of streamlining the file-load performance for the in-browser Less compiler is purely academical anyway. If you want increased page-load speed, you should precompile the Less files to plain old CSS files and serve those.

I essentially agree. I just wanted to open up the question and see if there was a quick switch to something that wouldn't trigger a parse but still maintain priority. Couldn't you just pass the root file into Less options? Then it wouldn't have to query the DOM, and it just retrieves it right away anyway. And the browser wouldn't try to parse it. Seems the most direct / efficient of all the options.

**Comment by rjgotten** [comment]
> Then it wouldn't have to query the DOM, and it just retrieves it right away anyway.

Actually, XHR gets scheduled at the very lowest priority of all possible request types by atleast Chrome/Blink and possibly others.

(Which is stupid really: you have a resource for which the website's own code is explicitly telling you that it needs it *now* and that resource gets ranked after images found by a speculative pre-parser...)

**Comment by whitelynx** [comment]
_@calvinjuarez wrote:_
> type="text/plain" would be valid, though not very semantic. type="text/less" is not valid, and maybe a little presumptuous, but it's semantic. Here's the text/x section of a list of media types for reference.

I'd suggest actually getting a media type registered. The `text/less` type would be the ideal solution, since LESS is an alternative format to CSS. (and the default type for `rel="stylesheet"` is `text/css`) It would probably also have the behavior you want in most cases. (i.e., if the page is _not_ in quirks mode) To quote [the HTML 5.1 spec section for the `stylesheet` link type](http://w3c.github.io/html/links.html#link-type-stylesheet):

> Once a resource has been obtained, if its Content-Type metadata is `text/css`, the user agent must run these steps:
> 1. Let _`element`_ be the link element that created the external resource link.
> 2. If _`element`_ has an associated CSS style sheet, remove the CSS style sheet in question.
> 3. If _`element`_ no longer creates an external resource link that contributes to the styling processing model, or if, since the resource in question was obtained, it has become appropriate to obtain it again (meaning this algorithm is about to be invoked again for a newly obtained resource), then abort these steps.
> _&lt;snip&gt;_

Basically, it seems that if the content type doesn't resolve to `text/css`, the link shouldn't be parsed as a stylesheet, even though its `rel` is set to `stylesheet`. Since this happens _after_ obtaining the resource, the fetching order should still be the one you want.

----

As far as the media type itself, you could always use `text/x-less` for now; though it seems to have been [deprecated by newer RFCs](https://tools.ietf.org/html/rfc6838#section-3.4), [the `x-` prefix](http://tools.ietf.org/html/rfc2046#section-6) is designed for "private" media type values, and as far as I know has pretty widespread support. Since it is deprecated, though, it would probably be best to pursue registration.

**Comment by tnga** [comment]
Because of `rel="stylesheet/less"`, browsers (as it was tested with firefox and chrome) seems to not emit the `onload` event. This can be normal considering `stylesheet/less` isn't actually a valid value for `rel`. Why not use perhaps a `data-sheetless` property instead and maintain default `rel` for now?
